### DISCUSSION TITLE
setting up labgym

### DISCUSSION THREAD

**Discussion by asmogul** [umyelab/LabGym#29] [Q&A]
What is the best setup to run labgym via computer (processor, RAM, hard drive)?

**Comment by yujiahu415** [comment]
The more advanced hardware of a computer, the faster running speed of LabGym. I think the 'best setup' depends on your need. For example, how fast do you want LabGym to run on your computer? An average laptop, for example, a one with a i5 processor, 8GB ram, 50GB free space of hard drive can run LabGym smoothly. But if you have a laptop with a i7 processor, 16GB ram, 100GB free space of hard drive, LabGym can run faster. So probability I can provide some idea about the 'minimum' requirement to run LabGym, instead of 'best'. But this 'minimum' requirement still varies based on users' need, the complexity of data, etc. So I would suggest, for a common scenario, that you have a computer with a processor of at least a i3 level, ram of at least 8GB, free hard drive space of at least 50GB to run LabGym smoothly.

**Reply by yujiahu415** [reply]
That being said, even if you have a computer that has lower-level hardware that what was aforementioned as ‘minimum’, LabGym can still run with a good speed if you choose to, for example, downsize the frame size of the videos, or a simpler Categorizer for behavior identification.

**Reply by asmogul** [reply]
Okay, thanks! Regarding frame size, how small a frame size can I go to optimize speed while not jeopardizing accuracy?

**Reply by yujiahu415** [reply]
That depends. A general rule is that, 1) after downsizing, the animal behaviors can still be identified by a human observer; 2) the estimated animal area after downsizing is still larger than the input size of the Categorizer. For example, if you are analyzing only one animal in a video and the animal occupies roughly 1/10 of the filed of view (frame), you may try to downsize the frame to roughly 320 x 320 if you still can see the animal behavior clearly. And use a Categorizer of input size as high as 96 x 96. But to increase speed, you may also use smaller input size of Categorizer, for example, 32 x 32, or even smaller. This is just an example to give you a brief idea. Just make sure that the estimated animal size after downsizing (320 x 320 x 1/10) is always larger than the input size of a Categorizer and you still see the behavior clearly in the video, then you may downsizing the video as much as possible to maximize the processing speed.

**Comment by asmogul** [comment]
I keep getting this error when trying to train categorizers with generated behavior examples:

ValueError: y has 0 samples: array([], dtype=float64)


Do you have any tips to fix this?

**Reply by yujiahu415** [reply]
To recap this error, I need more information from you. For example, 1. what are the behavior names / folder names for the training examples; 2. can you show me a screenshot of the behavior examples in the folder you used for training (the folder after the preparation). I need to see their file names. If you prefer to provide these info via email, here is my email: henryhu@umich.edu

**Reply by asmogul** [reply]
thank you! i am trying to download v1.9 because i think it was a problem with the pre-processing, but I am having trouble downloading the newer version

**Comment by asmogul** [comment]
When i try to install LabGym, 1.8 is automatically installed, however I would like version 1.9 for the preprocessing controls. Do you know how I can do this?

**Reply by yujiahu415** [reply]
If you have already installed LabGym v1.8 and want to upgrade it, you may use 'python3 -m pip install --upgrade LabGym'. If you have problem with installing v1.9, please show me the error message, or ideally the screenshot of your terminal / cmd which contains the necessary information I need to know so that I can see what is the problem and help you to address it. Thank you!

**Comment by asmogul** [comment]
Here is the error message I get when i try to update labgym:

Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [29 lines of output]
      Traceback (most recent call last):
        File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 156, in prepare_metadata_for_build_wheel
          hook = backend.prepare_metadata_for_build_wheel
      AttributeError: module 'sipbuild.api' has no attribute 'prepare_metadata_for_build_wheel

And this at the end:

 note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

**Reply by yujiahu415** [reply]
I guess the issue occurred when installing 'pytq5' but not sure about this because the error message you provided didn't tell me in which step this error occurred. But a quick fix might be upgrading your 'setuptools', 'wheel' and 'pip' and try upgrading LabGym again. You can run 'python3 -m pip install --upgrade setuptools wheel pip', after that run 'python3 -m pip install --upgrade LabGym'. If you still have this issue, see this: https://stackoverflow.com/questions/70961915/error-while-installing-pytq5-with-pip-preparing-metadata-pyproject-toml-did-n. This is a solution for the error when installing pytq5, and the latter part of the error message is similar to what you provided here. You can try their suggestions and let me know if it still doesn't solve the issue.

**Reply by yujiahu415** [reply]
A new version is out. I made it not relying on pytq5 so it should address your previous issue. You can upgrade LabGym.

**Reply by asmogul** [reply]
Thank you so much! I got it to work now. I am now having issues with sorting behavior examples. The behavior example generation is working but then when I try to sort the behavior examples nothing ends up being put in the new sorted folder. Then if i try to train the categorizer it says there is no data. Thoughts on how to fix this? Thanks!
